### Task 1 
##### In a Jupyter Notebook (.ipynb) file, use pandas to load the data from the JSON dataset into a DataFrame. Explore the data. 



In [3]:
# Import dependencies 

import pandas as pd
import json
from pandas.io.json import json_normalize

In [4]:
# Loading json file into a DataFrame

data = pd.read_json('dkdata.json')
data.head()

data
0  {'payer_name': 'AZARBAYCAN METANOL KOMPANI MMC...
1  {'payer_name': 'LCM ALLIANCE LLP', 'payer_juri...
2  {'payer_name': 'SKN ELECTRICAL SERVICES LIMITE...
3  {'payer_name': 'METASTAR INVEST LLP', 'payer_j...
4  {'payer_name': 'METASTAR INVEST LLP', 'payer_j...

In [5]:
# Checking number of rows 

data.shape

(16940, 1)

In [6]:
# Underlying numpy arrays to see what columns to expect 

data.values

array([[{'payer_name': 'AZARBAYCAN METANOL KOMPANI MMC', 'payer_jurisdiction': 'AZ', 'payer_account': '33817018409333311204', 'source_file': 'pdf/LCM ALLIANCE Account statement 30.06.12-31.12.14.xml', 'amount_orig': 535470.0, 'id': 6049, 'beneficiary_type': 'Company', 'beneficiary_core': True, 'amount_orig_currency': 'USD', 'beneficiary_name': 'LCM ALLIANCE LLP', 'beneficiary_jurisdiction': 'GB', 'investigation': 'az', 'beneficiary_bank_country': 'EE', 'beneficiary_name_norm': 'LCM ALLIANCE LLP', 'payer_core': False, 'beneficiary_account': 'EE27 3300 3335 0561 0002', 'purpose': '1206295100052180 OCT4121800021 ADVANCE PAYM FOR THE INST RUMENTATIO N AND CONTROL CABLES BOM IN ACC WIT H THE CON TR DD 25.06.2012', 'date': '2012-06-30', 'amount_usd': 535470, 'amount_eur': '$431,762.31', 'payer_type': 'Company', 'payer_name_norm': 'AZARBAYCAN METANOL KOMPANI MMC', 'payer_bank_country': '33'}],
       [{'payer_name': 'LCM ALLIANCE LLP', 'payer_jurisdiction': 'GB', 'payer_account': 'EE27 3300 3

In [7]:
# Unpacking nested column into a DataFrame utilizing json_normalize

with open('dkdata.json') as f:
    d = json.load(f)
    
dk_data = json_normalize(d['data'])
dk_data

amount_eur  amount_orig amount_orig_currency  amount_usd  \
0        $431,762.31    535470.00                  USD      535470   
1        $423,688.44   -535000.00                  USD      535000   
2         $71,698.53     90535.19                  USD       90536   
3         $71,686.50    -90520.00                  USD       90520   
4             $47.52       -60.00                  USD          60   
5        $189,352.15    189352.15                  EUR      239099   
6        $178,859.87    225850.00                  USD      225850   
7        $103,200.00   -103200.00                  EUR      130313   
8         $94,000.00    -94000.00                  EUR      118696   
9         $63,355.28    -80000.00                  USD       80000   
10        $55,435.87    -70000.00                  USD       70000   
11        $47,516.46    -60000.00                  USD       60000   
12        $43,556.76    -55000.00                  USD       55000   
13        $41,972.87    -53000.00                  USD       53000   
14        $29,250.34    -36935.00                  USD       36935   
15        $27,022.61    -34122.00                  USD       34122   
16         $3,567.69     -4505.00                  USD        4505   
17         $4,414.00     -4414.00                  EUR        5574   
18           $791.94     -1000.00                  USD        1000   
19            $22.91        28.93                  USD          29   
20       $118,791.15    150000.00                  USD      150000   
21       $118,791.15    150000.00                  USD      150000   
22       $125,786.00   -158832.54                  USD      158833   
23       $150,000.00   -150000.00                  EUR      189409   
24       $145,150.00   -145150.00                  EUR      183284   
25        $87,984.65   -111100.00                  USD      111100   
26        $72,400.00    -72400.00                  EUR       91421   
27        $68,366.00    -68366.00                  EUR       86328   
28        $25,342.11    -32000.00                  USD       32000   
29        $18,598.73    -23485.00                  USD       23485   
...              ...          ...                  ...         ...   
16910    $105,000.00   -105000.00                  EUR      130967   
16911     $91,000.00    -91000.00                  EUR      113505   
16912     $60,210.00    -75100.00                  USD       75100   
16913     $66,338.00    -66338.00                  EUR       82744   
16914     $64,825.00    -64825.00                  EUR       80857   
16915     $55,007.00    -55007.00                  EUR       68611   
16916     $40,888.28    -51000.00                  USD       51000   
16917     $32,069.24    -40000.00                  USD       40000   
16918     $25,000.00    -25000.00                  EUR       31183   
16919     $20,043.28    -25000.00                  USD       25000   
16920     $21,839.20    -21839.20                  EUR       27241   
16921     $16,034.62    -20000.00                  USD       20000   
16922      $7,668.05     -9564.37                  USD        9565   
16923      $7,040.00     -7040.00                  EUR        8781   
16924      $6,876.00     -6876.00                  EUR        8577   
16925      $1,892.00     -1892.00                  EUR        2360   
16926      $1,547.34      1930.00                  USD        1930   
16927    $300,000.00    300000.00                  EUR      374191   
16928  $4,409,477.61   5499946.50                  USD     5499947   
16929  $4,712,772.44   5878246.50                  USD     5878247   
16930    $300,000.00   -300000.00                  EUR      374191   
16931    $230,000.00   -230000.00                  EUR      286880   
16932    $134,600.00   -134600.00                  EUR      167887   
16933    $962,077.20  -1200000.00                  USD     1200000   
16934    $617,332.87   -770000.00                  USD      770000   
16935     $80,173.10

### Basic Analysis 

Now that we were able to properly load all the content of the file into a DataFrame we are going to look for basic informations like, duplicate names, missing values, account numbers etc... 

In [8]:
# Checking for any missing value and summing them up

dk_data.isnull().sum()

amount_eur                    0
amount_orig                   0
amount_orig_currency          0
amount_usd                    0
beneficiary_account           4
beneficiary_bank_country      4
beneficiary_core              0
beneficiary_jurisdiction      0
beneficiary_name              0
beneficiary_name_norm         0
beneficiary_type              0
date                          0
id                            0
investigation                 0
payer_account               115
payer_bank_country          115
payer_core                    0
payer_jurisdiction            0
payer_name                    0
payer_name_norm               0
payer_type                    0
purpose                       0
source_file                   0
dtype: int64

On the above cell we have learned that there are 115 payers's bank account missing, consequentely 115 payer's bank country missing. Also, we noticed that there are 4 beneficiary bank account missing, consequentely 4 beneficiary bank country missing. I'm still not sure what that means, however, it raises questions as I wonder if they are somehow corelated. 

In [22]:
# Pulling data that only contains missing information from initial 
# dataframe

dk_dataNull = dk_data[dk_data.isnull().any(axis=1)]
dk_dataNull

amount_eur  amount_orig amount_orig_currency  amount_usd  \
67     $132,789.00    132789.00                  EUR      167676   
121     $39,567.35     49962.50                  USD       49963   
306      $1,468.05     -1800.00                  USD        1800   
308     $66,871.29     66871.29                  EUR       81992   
451     $30,111.44     36920.00                  USD       36920   
509      $3,262.34      4000.00                  USD        4000   
603     $12,618.69     10000.00                  GBP       15872   
1210       $178.33       178.33                  EUR         230   
1211     $1,242.04      1242.04                  EUR        1601   
2923     $1,844.77      1500.00                  GBP        2437   
3084     $8,790.90     12000.00                  USD       12000   
3508    $45,732.46     62427.00                  USD       62427   
3510    $73,215.40     99942.53                  USD       99943   
3598     $2,131.43     -2909.50                  USD        2910   
3677    $14,483.01     19770.00                  USD       19770   
4229   $500,000.00    500000.00                  EUR      651330   
4441   $274,862.00    274862.00                  EUR      358052   
4823   $360,438.00    360438.00                  EUR      462652   
5073   $475,578.00    475578.00                  EUR      610443   
5124   $103,053.00    103053.00                  EUR      132277   
5126   $122,920.00    122920.00                  EUR      157778   
5590   $154,379.15    198158.00                  USD      198158   
6232    $46,861.45   1927646.00                  RUB       61781   
6234    $56,537.53   2325672.00                  RUB       74538   
6236    $88,271.74   3631059.00                  RUB      116375   
6237    $90,383.84   3717940.00                  RUB      119159   
6385   $155,167.34    204567.00                  USD      204567   
6493   $204,874.00    204874.00                  EUR      270099   
6636        $20.36        26.84                  USD          27   
6643   $126,200.00    126200.00                  EUR      166378   
...            ...          ...                  ...         ...   
12014   $55,541.54     75482.00                  USD       75482   
12160   $39,950.00     39950.00                  EUR       54293   
12383   $50,869.90     69975.33                  USD       69976   
12572    $8,152.68     10000.00                  CHF       11215   
12599   $39,950.00     39950.00                  EUR       53887   
12829   $11,120.60     15000.00                  USD       15000   
12872       $38.65        53.35                  USD          54   
12873    $4,075.00      4075.00                  EUR        5625   
12900       $38.65        53.35                  USD          54   
12901    $3,172.03      4378.00                  USD        4378   
12940  $211,411.00    211411.00                  EUR      291787   
12941  $399,978.00    399978.00                  EUR      552045   
12943  $912,012.74   1258749.00                  USD     1258749   
12946   $10,004.43     13808.00                  USD       13808   
12948  $181,097.80    249949.00                  USD      249949   
12949  $673,784.32    929949.00                  USD      929949   
12974       $38.99        53.81                  USD          54   
12975    $4,000.00      4000.00                  EUR        5521   
13274   $82,539.00     82539.00                  EUR      113853   
13536    $7,135.00      7135.00                  EUR        9896   
14156    $8,238.26     11227.00                  USD       11227   
14317   $46,249.25     63265.00                  USD       63265   
15343   $53,315.99     70000.00                  USD       70000   
15613   $93,403.52    122632.00                  USD      122632   
15993  $131,950.00    131950.00                  EUR      166510   
16089  $349,950.00    349950.00                  EUR      441607   
16171  $292,950.00    292950.00                  EUR      366914

In [34]:
# Count of the unique payers name 
# Name of the unique payers name 

print(dk_dataNull.payer_name.nunique())
dk_dataNull.payer_name.unique()

70


array(['DEUTDEFFXXX', 'BKCHCNBJ', 'METASTAR INVEST LLP', 'DENITRIS',
       'TDOMUS33', 'BARCGB22', 'JOHNS HOPKINS MEDICINE', 'NORSDE71',
       'TD CANADA TRUST', 'ROYCCAT2X', 'HASEHKHH', 'OWHBDEFF', 'PANO ASC',
       'TRADEMAX GENERAL TRADING CO.LTD', 'HYIBLI22', 'INN3018016381',
       'INN3016055642', 'INN5918000980', 'INN5906109648',
       'ALPS VENTURES LIMITED', 'SABA CARS GERMANY GMBH',
       'MR RAHEB GENERAL TRADING', 'INN3016043171', 'INN3017030922',
       'In favour JSC Freight One', 'SHANGHAI PUDONG DEV BANK',
       'SBERBANK OF RUSSIA', 'INN6234018156',
       'ALGHARABALLY NATIONAL TRADING',
       'UNTEL KABLOLARI SANAYI VE TICARET', 'KEMIRA OYJ',
       'UCGE ELEKTONIK CIH.BILG.VE MAG.EKP.', 'GENERAL TURBO S.A',
       'MARAK CO LIMITED', 'PACIFIC LAMPS AND ELECTRICAL',
       'QIN YI TRADING CO.,LIMITED', 'HAENG SUNG ELECTRONIC CO',
       'YIWU HONGCHENG TRADING CO.,LTD', 'SANDLAND OVERSEAS',
       'XINJIANG KAIMEI INTERNATIONAL', 'RUSSIAN AGRICULTURAL BANK',
 

In [32]:
# Count of the unique beneficiary name 
# Name of the unique beneficiary name 

print(dk_dataNull.beneficiary_name.nunique())
dk_dataNull.beneficiary_name.unique()

7


array(['METASTAR INVEST LLP', 'CONTINENTAL ENGINES', 'LCM ALLIANCE LLP',
       'MICROCOPY', 'POLUX MANAGEMENT LP', 'HILUX SERVICES LP',
       'TURBILLION LLC'], dtype=object)

In [37]:
# Exporting dk_data dataframe into a csv file

dk_data.to_csv('dk_data.csv', index=False)